echopop/utils/batch_load.py

```diff
@@ -0,0 +1,608 @@
+        pattern = "|".join(
+            [r"(?<!-)\b{}\b".format(re.escape(name.lower())) for name in region_names[key]]
+        )
+        # ---- Apply filter to regions dataframe
+        new_region_filtered = new_region_df[
+            new_region_df["region_class"].str.contains(pattern, case=False, regex=True)
+        ]
+        # ---- Concatenate to the initialized/pre-existing DataFrame
+        region_df = pd.concat([region_df, new_region_filtered], ignore_index=True)
+
+    # return
+    return region_df
+
+
+def get_haul_transect_key(configuration_dict: dict, root_dir: str):
+    """
+    Get the haul-to-transect KS stratification key.
+    """
+
+    # Flatten the configuration settings
+    flat_table = pd.json_normalize(configuration_dict)
+
+    # Get the configuration specific to KS strata
+    strata_config = flat_table.filter(regex=r"\bstrata\b")
+
+    # Get the target file and sheet
+    # ---- Get filename
+    file = strata_config.filter(regex="filename").values.flatten()
+    # ---- Get sheetname
+    sheet = strata_config.filter(regex="sheetname").values.flatten()
+
+    # Get configuration map settings specific to the KS strata
+    # ---- Dictionary to DataFrame
+    CONFIG_DF = pd.DataFrame(CONFIG_MAP)
+    # ---- Find 'strata'
+    strata_config = CONFIG_DF.loc["strata"].dropna()
+    # ---- Get the first valid index
+    strata_reference = strata_config[strata_config.first_valid_index()]
+
+    # Create filepath
+    # ---- Create filepath
+    haul_key_path = Path(root_dir) / file[0]
+    # ---- Validate
+    if not haul_key_path.exists():
+        raise FileExistsError(f"The haul-to-transect file {{{str(haul_key_path)}}} does not exist!")
+
+    # Read in data
+    haul_key = pd.read_excel(haul_key_path, sheet_name=sheet[0])
+    # ---- Retain only the necessary columns
+    haul_key_filtered = haul_key.filter(strata_reference.keys())
+    # ---- Assign datatypes
+    haul_key_filtered = haul_key_filtered.astype(strata_reference, errors="ignore")
+
+    # Return output
+    return haul_key_filtered
+
+
+def filter_export_regions(
+    transect_data: pd.DataFrame,
+    region_filter: Union[str, list[str]],
+    index_variable: Union[str, List[str]] = ["transect_num", "interval"],
+    unique_region_id: str = "region_id",
+    region_class_column: str = "region_class",
+    impute_regions: bool = True,
+):
+
+    # Check if region column is present if region filter is defined
+    if region_class_column not in transect_data.columns:
+        raise ValueError(
+            f"The defined `region_class_column` ({region_class_column}) used for applying "
+            f"`region_filter`does not exist!"
+        )
+    # ---- Convert to list, if needed
+    if isinstance(region_filter, str):
+        region_filter = list(region_filter)
+    elif not isinstance(region_filter, list):
+        raise TypeError(
+            f"The defined `region_filter` ({region_filter}) must be either a `str` or `list`."
+        )
```

Comment:
Seems that the isinstance check for both list and str should happen first, and then the conversion from str to list happens if necessary.

Reply:
Same with the `index_variable` check below.